### ISSUE TITLE
Use syslog-ng with SQLite database 3.24.0

### ISSUE THREAD

**Issue by Wernfried** [balabit/syslog-ng#2424]
I am running syslog-ng version 3.7.3 and logs are written to SQLite database.

I created a trigger in SQLite DB which uses the new [UPSERT](https://www.sqlite.org/lang_UPSERT.html) command. UPSERT syntax was added to SQLite with version 3.24.0 (2018-06-04).

Now I get an error:
`Error running SQL query; type='sqlite3', host='', port='', user='', database='/var/lib/mediation/sessionlog-2-46.db', error='11: malformed database schema (Action_close) - near "ON": syntax error', query='PRAGMA journal_mode = MEMORY'`

I assume the syslog-ng driver does not support the new UPSERT command in SQLite. (I get a similar error message when I try to run this SQLite database-file with an old version of sqlite3 application)

Would it be sufficient to update to the latest release of syslog-ng to make it running? Or are there any third-party drivers which might not be updated yet.

Best Regards
Wernfried

**Comment by Kokan** [comment]
The `libdbi` is used to connect to sql databases, which has `libdbi-drivers` (that lib has the sqlite specific connector). You could check if there is update for those.

But based on your comment, I think you probably has old `sqlite` lib linked against the above mentioned 3rd party libs, and that is causing the issue.
As the trigger should be transparent from syslog-ng point of view.
Did you have other trigger before this one ? (just to verify if syslog-ng and the trigger is compatible)

**Comment by Wernfried** [comment]
Yes, application works fine with other triggers which do not apply UPSERT statement. So, it is fairly obvious that error is caused by this new function.

Does syslog-ng uses the this libdbi drivers: https://sourceforge.net/projects/libdbi-drivers/ ?

In this case I would not have much hope as it was not updated since 2013: 
https://sourceforge.net/projects/libdbi-drivers/files/libdbi-drivers/ 

Name 	Modified
libdbi-drivers-0.9.0 	2013-03-12 		
libdbi-drivers-0.8.3-1 	2008-03-03 		
libdbi-drivers-0.8.3 	2008-02-06

**Comment by Kokan** [comment]
Yes it is.
But the day might be saved, even with dinosaurs, as my suggestion is not to update `libdbi` or `libdbi-drivers` but to update the `sqlite`.

As my tipp (no proof of course) that actually what is blocking you is the `libsqlite3`, which is linked to the `libdbdsqlite3.so` (the libdbi-driver connector for sqlite3).

Could you checked if the `libdbdsqlite3.so` uses the sqlite3 lib with correct version or not ?
If not, maybe you only have to rebuild the `libdbdsqlite3.so` agains the updated sqlite3 lib, and it could still work.

**Comment by Wernfried** [comment]
I am not so familiar with deploying applications on Linux but let's try.

Currently I have this one on my machine:

```
lrwxrwxrwx. 1 root root     19 Jul 22  2016 /usr/lib64/libsqlite3.so.0 -> libsqlite3.so.0.8.6
-rwxr-xr-x. 1 root root 753232 Jul 23  2015 /usr/lib64/libsqlite3.so.0.8.6

```

I compiled the latest source "sqlite-autoconf-3250300" of sqlite and I get this:

```
lrwxrwxrwx. 1 domscheit domain users      19 Nov 23 11:24 /home/domscheit/sqlite-autoconf-3250300/.libs/libsqlite3.so -> libsqlite3.so.0.8.6
lrwxrwxrwx. 1 domscheit domain users      19 Nov 23 11:24 /home/domscheit/sqlite-autoconf-3250300/.libs/libsqlite3.so.0 -> libsqlite3.so.0.8.6
-rwxr-xr-x. 1 domscheit domain users 5192741 Nov 23 11:24 /home/domscheit/sqlite-autoconf-3250300/.libs/libsqlite3.so.0.8.6
```
Version seems to be the same (0.8.6) but my compiled file is 6 times bigger than the existing one. Does it mean I have to replace `/usr/lib64/libsqlite3.so.0.8.6` with the new one?

Thank you

**Comment by Kokan** [comment]
The file size difference probably because of debug symbols. (if you want you can verify: `file [sopath]" should print `with debug_info, not stripped`).

My suggestion was to check the `libdbi-driver` (and possibly recompile).
You can find it `find / -name libdbdsqlite3.so`, and check it via `ldd [path found before]`, and the list should have the `libsqlite3` referenced, this should give you a hint which version of `libsqlite3` is used for `libdbdsqlite3`.
If it is the same `0.8.6` that works for you, recompiling do not make sense, and the issue is probably in that library.
If the version is older, maybe you could try to recompile the `libdbi-drivers` with the newer `libsqlite3` in a hope of solving your issue.

Sadly if it is the fault of `libdbi`, there is not much to do. I've tried to contribute fix in the past, no reaction at all.
On the long term it is planned to replace the sql lib with something, but I do not know who and when is going to do it.

**Comment by Wernfried** [comment]
That's what I did:

```
[mediation@t-mipmed-as-01 ~]$ find / -name "*libsqlite3*" 2> /dev/null
/usr/lib64/libsqlite3.so.0
/usr/lib64/libsqlite3.so.0.8.6
/usr/openv/pdde/pdopensource/lib/libsqlite3.so
/usr/openv/pdde/pdopensource/lib/libsqlite3.so.0
/usr/openv/pdde/pdopensource/lib/libsqlite3.so.0.8.6
[mediation@t-mipmed-as-01 ~]$ ldd /usr/lib64/libsqlite3.so.0 /usr/lib64/libsqlite3.so.0.8.6
/usr/lib64/libsqlite3.so.0:
        linux-vdso.so.1 =>  (0x00007ffc0a9a9000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f5695001000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f5694de5000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f5694a18000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f56954ba000)
/usr/lib64/libsqlite3.so.0.8.6:
        linux-vdso.so.1 =>  (0x00007ffff2fbf000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f865c85b000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f865c63f000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f865c272000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f865cd14000)
[mediation@t-mipmed-as-01 ~]$ file /usr/lib64/libsqlite3.so.0.8.6
/usr/lib64/libsqlite3.so.0.8.6: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=39ae83139767c2121e2aba701b03a23e98def484, stripped
```

And my new one:
```
[domscheit@d-dhcplog-as-01 ~]$ file /home/domscheit/sqlite-autoconf-3250300/.libs/libsqlite3.so.0.8.6
/home/domscheit/sqlite-autoconf-3250300/.libs/libsqlite3.so.0.8.6: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=62878f59956d3c201dc0231b20ec4b7e50462c14, not stripped

[domscheit@d-dhcplog-as-01 ~]$ ldd /home/domscheit/sqlite-autoconf-3250300/.libs/libsqlite3.so.0.8.6
        linux-vdso.so.1 =>  (0x00007fff773f5000)
        libm.so.6 => /lib64/libm.so.6 (0x00007efd85ea6000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007efd85ca1000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007efd85a85000)
        libc.so.6 => /lib64/libc.so.6 (0x00007efd856c3000)
        /lib64/ld-linux-x86-64.so.2 (0x00007efd864ba000)
```

I compiled latest version of libdbi-drivers but I don't managed to get anything newer than libsqlite3.so.0.8.6

Locks like for the moment I have to live with this limitation or I am not smart enough to make a proper compile.

**Comment by Kokan** [comment]
My os does have older version of sqlite3, so I picked the source from https://sqlite.org/download.html
compiled and installed it into `/tmp/install`.
I created a database with `/tmp/insta//bin/sqlite3` and after tried with the one on the system, and it wrote the same error as yours.
Started syslog-ng with preloading the newly compiled sqlite3, and it seemed to work for me.
(`LD_PRELOAD=/tmp/install/lib/libsqlite3.so`)

What OS are you using, how did you installed syslog-ng ? Because I think you have at least two sqlit3 installed, that causing the issue. You should use one, and it should work.

**Comment by Wernfried** [comment]
I just did a "quick and dirty" test on my development machine with latest release 3.18 and it seems to work. Would be great if a simple upgrade of syslog-ng would solve my issue. But I have to run more careful tests next week on test machine.
